### ISSUE TITLE
Multiple selection support for useComboBox

### ISSUE THREAD

**Issue by stephenh** [adobe/react-spectrum#2140]
# 🙋 Feature Request

Update useComboBox to support multiple selection.

## 🔦 Context

We're building a type-ahead `SelectField` / `MultiSelectField` components, both on top of `useComboBox` and having to support multiple selections is of course making us do some hacky things, b/c it's not really supported.

I.e. we're not passing `selectedKey` to `useComboBox` b/c in a multiple selection scenario, the `selectedKey` could change, and so then odd things happen where `useComboBox` thinks its not controlled and/or the `selectedKey` !== `inputValue`, so it invokes various closing/reset behavior that we didn't mean to trigger.

Obviously this is on us for misusing `useComboBox` :-D , so just filing a request / wish list for it to get official multiple selection support. Thanks!

**Comment by devongovett** [comment]
We call this a tag field - a combination of a tag list and an input kinda like the "to" field in an email app. It's on the roadmap, but not sure when yet. We need to do some a11y research and testing to determine how best to implement it.

**Comment by stephenh** [comment]
@LFDanLu lazy ask, but is there a ticket created for the tag field that @devongovett mentioned? I'd like to follow it, so that I can see when react-aria lands support for it. Thanks!

**Comment by LFDanLu** [comment]
@stephenh I don't believe there is a ticket created for the tag field yet since it is still in its early germination phase unfortunately. https://github.com/adobe/react-spectrum/projects/1 is the best place to look regarding what is currently in flight.

**Comment by devongovett** [comment]
Going to keep this ticket open for tracking.

**Comment by ashiskumar-1999** [comment]
Hey @devongovett , Does it look like similar to this?
https://react-listbox.vercel.app/?path=/story/example-listbox--multiselect
https://www.telerik.com/kendo-react-ui/components/dropdowns/multiselect/

**Comment by devongovett** [comment]
More like the second one, the first is only a select not a combobox (no textfield). Think similar to an email entry field like in Gmail. Here's a screenshot of the design from Spectrum that we'd be basing the pattern on:

![tag-field_hero_desktop@2x_6md9yFlaUX9BDThTdZNbgC_1611704216000](https://user-images.githubusercontent.com/19409/130142554-0e80b8bf-c323-4de1-a102-f490e8659002.png)

**Comment by ashiskumar-1999** [comment]
@devongovett 
If this falls under priority, I can start working on it.

**Comment by ashiskumar-1999** [comment]
Hey, @devongovett As I move forward trying to contribute to this task, I'm wondering if this use-case requires us to build another independent component/hook or this functionality could be incorporated into any existing component?

**Comment by beamery-tomht** [comment]
Is there a way to achieve this behaviour with existing hooks in `react-stately` and `react-aria`?

I'm working on a custom component library according to a design system and using `react-aria` extensively (without spectrum) and a multi-select is a new requirement. It would be great if we could leverage react-aria for this without having to lean on another library or build something from scratch.

From what I can see, some combination of a SelectionManager and ListBox looks like the most suitable approach

**Comment by jamesonhill** [comment]
> Is there a way to achieve this behaviour with existing hooks in `react-stately` and `react-aria`?
> 
> I'm working on a custom component library according to a design system and using `react-aria` extensively (without spectrum) and a multi-select is a new requirement. It would be great if we could leverage react-aria for this without having to lean on another library or build something from scratch.
> 
> From what I can see, some combination of a SelectionManager and ListBox looks like the most suitable approach

This also matches my use case. Are we better off building something from scratch or is support for this use case on the near term roadmap?

**Comment by crazko** [comment]
> Is there a way to achieve this behaviour with existing hooks in `react-stately` and `react-aria`?
> 
> I'm working on a custom component library according to a design system and using `react-aria` extensively (without spectrum) and a multi-select is a new requirement. It would be great if we could leverage react-aria for this without having to lean on another library or build something from scratch.
> 
> From what I can see, some combination of a SelectionManager and ListBox looks like the most suitable approach

I was facing this problem regarding multiselect as well. Since [`@react-aria/select`](https://react-spectrum.adobe.com/react-aria/useSelect.html) and suitable stately hooks do not support multiple selection mode (and I totally get [why it is the case](https://github.com/adobe/react-spectrum/discussions/2218#discussioncomment-1159271)), I followed the [implementation of the Picker component](https://github.com/adobe/react-spectrum/blob/main/packages/@react-spectrum/picker/src/Picker.tsx), re-created hooks to support our needs (multiple selection, searching possibility, clear & select all buttons...) and built a new Select component with them.

So, instead of:

- [`useSelectState`](https://github.com/adobe/react-spectrum/blob/main/packages/@react-spectrum/picker/src/Picker.tsx#L66) there's our new `useMultiSelectState`
  - inside it, [`useSingleSelectListState`](https://github.com/adobe/react-spectrum/blob/main/packages/@react-stately/select/src/useSelectState.ts#L33), there's `useMultiSelectListState`
- [`useSelect`](https://github.com/adobe/react-spectrum/blob/main/packages/@react-spectrum/picker/src/Picker.tsx#L79) there's `useMultiSelect`

Basically it means a lot of copy-pasting and playing with properties of hooks underneath. Once I became more familiar with the original implementation I found it very flexible and customizable.

I cannot share the whole source code, but I'm happy to answer any questions regarding that.

**Comment by snowystinger** [comment]
Just an FYI that we are also on our way to building what was described in https://github.com/adobe/react-spectrum/issues/2140#issuecomment-902237781 This is the PR where TagGroup work has been started
https://github.com/adobe/react-spectrum/pull/1922
Once we have that, we'll be able to consider joining it with a combobox or other components to handle multi select

**Comment by donaldpipowitch** [comment]
@crazko are your multi select compatible hooks public by any chance?

**Comment by crazko** [comment]
@donaldpipowitch let's check example at https://codesandbox.io/s/spectrum-multiselect-9be866-9be866. It is a simplified version of our `Select` component implementation with support for both single & multiple selection mode.

There's a bug though, which I was not able to find a fix for so far; it is not present in our library. But should give enough input for how this may work.

**Comment by donaldpipowitch** [comment]
Thank you very much!

**Comment by npearson72** [comment]
Wondering if it will be possible to update the input (or contenteditable element) with a mix of values. My use case is an omni-search box that allows users to type random search strings, but as they type a menu can appear to offer them selections for tags, categories, and other meta data fields. If they make a selection, I want them styled selected (ex: tag) to appear in the input field where their cursor was, allowing them to continue typing/selecting as they wish.

**Comment by snowystinger** [comment]
If you are using our hooks to make your combobox, this should already be possible. I wouldn't actually update the input for mixed anything, but you could render anything you wanted before or after the combobox, then just make it visually look like it's part of the input.

**Comment by npearson72** [comment]
@snowystinger Thanks for the hint. But from what you're describing I'm struggling to understand how a user would interact with the data they've entered.

For example, say they enter the following:

The quick `brown` fox jumped `over` the `lazy` dog

But now they want to change the word "jumped" to "ran". How would it be possible if this was rendered outside of the input field (or contenteditable element)?

**Comment by snowystinger** [comment]
hmmm have you tried tokenizing to words, and each word is a combobox? because it doesn't seem like the whole thing is one combobox.

I'm not sure the accessibility implications of this approach, but I don't think it fits combobox otherwise at all. This is an, you're off the map, situation.

Keep in mind, this https://www.w3.org/WAI/ARIA/apg/patterns/combobox/ is the aria pattern we have implemented.

**Comment by npearson72** [comment]
@snowystinger Got it. Thanks for the explanation. I think you're right in that it might be some hybrid pattern. Appreciate your feedback.

**Comment by andresgutgon** [comment]
Hi, yesterday I was also looking for this. I was able to do an autocomplete + tags using 
1. [useKeyboard](https://react-spectrum.adobe.com/react-aria/useKeyboard.html)
2. [useListBox](https://react-spectrum.adobe.com/react-aria/useListBox.html)
3. [useListState](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/list/src/useListState.ts)

![image](https://user-images.githubusercontent.com/49499/221219738-5c061399-320c-4d3c-97e8-e0aadf277a59.png)

Basically I reimplemented `useComboBox` with those primitives but configuring `useListState` as `multiple` to be able to handle multiple selection 🎉 

![image](https://user-images.githubusercontent.com/49499/221220406-5c6ba72a-690a-43ee-97ab-06f90c997fce.png)

**Comment by wporoslo** [comment]
@andresgutgon Could you kindly share your implementation? I'm struggling with this

**Comment by andresgutgon** [comment]
It's still in the oven, but when I have time I'll share a sandbox with the parts.

**Comment by andresgutgon** [comment]
@wporoslo sorry for the delay. I made a very ugly version of the one we built
![image](https://user-images.githubusercontent.com/49499/222378543-2f38f643-3a6a-4dcf-94e6-68eecc42bb15.png)

But the parts necessary to do a combobox + multiple selection are there
https://codesandbox.io/s/sleepy-keller-qjygke?file=/src/App.tsx

**Comment by wporoslo** [comment]
> @wporoslo sorry for the delay. I made a very ugly version of the one we built
thank you so much!

**Comment by yarastqt** [comment]
Is there any status on this issue? This is a very important feature that we are waiting for 😔

**Comment by snowystinger** [comment]
> Is there any status on this issue? This is a very important feature that we are waiting for 😔

Please see the discussion above, it should answer any questions you have as well as how to build your own. If the ask is specifically for a React Spectrum TagField component, that should be moved to a different place. useCombobox is a specific aria pattern that will not be supporting multiple selection on its own.

**Comment by amauryfischer** [comment]
Now that Taggroup is released i would love to see a Tagfield / comboBox with multiple selection.
Like "Autocomplete" in material ui.

**Comment by tekno0ryder** [comment]
NextUI just dropped their [select component](https://nextui.org/docs/components/select) and it supports multi selection 👏🏽

**Comment by huongdevvn** [comment]
> @donaldpipowitch let's check example at https://codesandbox.io/s/spectrum-multiselect-9be866-9be866. It is a simplified version of our `Select` component implementation with support for both single & multiple selection mode.
> 
> There's a bug though, which I was not able to find a fix for so far; it is not present in our library. But should give enough input for how this may work.

I fixed a bug in this example by adding `placeholder` like:
![Screenshot 2024-02-01 at 14 52 21](https://github.com/adobe/react-spectrum/assets/45929865/9692cf21-a701-4c92-ae4c-795a437beac5)

**Comment by mryechkin** [comment]
> This project is using `react-aria` and has `multi-select` support
> 
> https://project44.github.io/manifest/inputs/mult-combobox
> 
> but I'm eager to see it in `react-aria-components`

Great find! I agree that it would be good to see it in RAC, but that's at least a good reference implementation if someone wants to build their own.

Here's the source code for anyone interested: https://github.com/project44/manifest/tree/main/packages/react/src/components/MultiCombobox

**Comment by deltasierra96** [comment]
> This project is using `react-aria` and has `multi-select` support
> 
> * https://project44.github.io/manifest/inputs/mult-combobox
> * https://pluralsh-design.web.app/?path=/story/combo-box--cluster-tags
> * https://pluralsh-design.web.app/?path=/story/select--default
> 
> but I'm eager to see it in `react-aria-components`

Thanks for the reference. I don't suppose you were able to replicate such a component with RAC?

**Comment by GermanJablo** [comment]
> Please see the discussion above, it should answer any questions you have as well as how to build your own. If the ask is specifically for a React Spectrum TagField component, that should be moved to a different place. useCombobox is a specific aria pattern that will not be supporting multiple selection on its own.

This comment left me confused because before you said that:

> Once we have that, we'll be able to consider joining it with a combobox or other components to handle multi select

If I understand correctly, this is planned, but it would not be a Combobox functionality but rather a new input type called TagField, is that correct?

If it's a name problem, perhaps this issue could be renamed instead of opening a new one :)

**Comment by snowystinger** [comment]
@GermanJablo your assessment is correct. I'm a little wary of changing the name of the issue because people do search for it a lot and if it said "TagField", I fear we'd get a lot of new issues opened.

The only reason for a new issue I feel would be for the React Spectrum implementation, since we'd be building and styling that.
I'm not sure how much needs to be done for RAC. I think it's fairly close already to being possible. Here's a real quick sketch of what it'd take.
https://stackblitz.com/edit/vitejs-vite-iycf5n?file=src%2FApp.tsx,package.json (styling would need updates to make it look all like one input, but is doable already)
Would need to hook up the labels to a group label probably through aria-labelledby. I'm sure there are some other things that need tweaking or adjustments. I think the bigger question will be if anything is actually missing from the components used that prevents this.

**Comment by GermanJablo** [comment]
Oh, I understand. This design pattern seems to rely a lot on CSS compared to the rest of the RACs.
Regarding your question, I don't know what might be missing from RAC in terms of structure or a11y, but perhaps the best way to find out is to try to make a styled example.

**Comment by ryo-manba** [comment]
Hi! I'm interested in this implementation and think having a specification similar to what's detailed in the following issues would be  helpful during development.

- https://github.com/adobe/react-spectrum/issues/2396
- https://github.com/adobe/react-spectrum/issues/2083

Currently, it seems there isn't a detailed spec for this component, and I'm eager to learn more about any research or preliminary considerations you might have at this point, even if it's just in note form.

For example, I'm curious about concepts like moving the cursor between tags, as mentioned in [Devon's post](https://x.com/devongovett/status/1765095544563597753).

**Comment by ryo-manba** [comment]
@snowystinger 
> I think the bigger question will be if anything is actually missing from the components used that prevents this.

There are a few problems with existing hook combinations that I'd like to introduce.

### 1. Unavailability of `isSelected`

The current `useComboBoxState` is designed exclusively for single selections, where `isSelected` is applied only to the last selected element. This limitation complicates styling based on selection states within a Popover. For example, the following code does not perform as expected.

```tsx
<ListBoxItem>
  {({isSelected}) => (
    <>
      {isSelected && <CheckmarkIcon />}
      Item
    </>
  )}
</ListBoxItem>
```

As a workaround, one could manage separate lists for selectable and selected items and reflect changes in real time using a map function. However, this could lead to unnecessary rendering, making it desirable to utilize `isSelected` effectively.

```tsx
const list = useListData({initialItems: items})
const selectedList = useListData()
const isSelected = (item) => typeof selectedList.getItem(item.id) !== 'undefined';

<ListBox>
  {list.items.map((item) => (
      <ListBoxItem id={item.id} key={item.id}>
        {isSelected && <CheckmarkIcon />}
        Item
      </ListBoxItem>
  ))}
</ListBox>;
```

### 2. Non-Reflective `aria-selected` in Popover Choices

Specifying `selectionMode` and `selectedKeys` in a `ListBox` does not reflect `aria-selected` or `aria-multiselectable` due to the use of the ComboBox context.

```tsx
<ListBox selectionMode="multiple" selectedKeys={selectedKeys}>
```

Here are some examples where aria and data attributes are set based on the ComboBox context:

- [Processing `isSelected` in `useOption`](https://github.com/adobe/react-spectrum/blob/72775d272be04cfa32d26ec821bc68513d92a72b/packages/%40react-aria/listbox/src/useOption.ts#L97)
- [Setting of `aria-selected` in `useOption`](https://github.com/adobe/react-spectrum/blob/72775d272be04cfa32d26ec821bc68513d92a72b/packages/%40react-aria/listbox/src/useOption.ts#L109)
- [Setting of `aria-multiselectable` in `useListBox`](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/listbox/src/useListBox.ts#L116-L118)
- [Setting of `data-selection-mode` in `ListBox`](https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/src/ListBox.tsx#L443)

To address these issues, it might be advisable to create dedicated hooks for the TagField, rather than merely relying on the combination of existing hooks.

**Comment by snowystinger** [comment]
Thanks for having a think about this!

1. I think this should be a non-issue. We were to remove the selected item from the combobox entirely and place it into the taggroup. Assistive technology wouldn't know how to announce multiple selection in a combobox.
https://www.w3.org/WAI/ARIA/apg/patterns/combobox/
> The value must be one of a predefined set of allowed values ... or an arbitrary value

2. See #1, theoretically in this case the combobox will never actually have a selected value, since it would be removed and placed in the taggroup

Let us know if you have more questions or think of something else.

**Comment by ryo-manba** [comment]
@snowystinger 
> We were to remove the selected item from the combobox entirely and place it into the taggroup.

Is it correct to understand that the drop-down list of the ComboBox will not support use cases where selected items are displayed within the list?

**Comment by chrisFFL** [comment]
Hello, Chris ***@***.*** is no longer with our company. Can you
please remove him from your system?

Thank you,
Olya

**Comment by snowystinger** [comment]
Yes, that's correct

**Comment by 6thpath** [comment]
Hi @snowystinger, is there any chance some hook/context like `useCollectionDocument`, `CollectionDocumentContext` of `Collection` could be exported through 
`react-aria-components`, I almost done on creating a select-multiple component except that list-box-item won't show due to these api are not exported
<img width="261" alt="Screenshot 2024-04-30 at 00 42 43" src="https://github.com/adobe/react-spectrum/assets/22215743/41d8ee92-ac8c-44d0-9b4f-d26531ef326f">

**Comment by sadeghbarati** [comment]
Hey @6thpath, could you please share your code on StackBlitz?

We can manually patch `react-aria-components` for now to export `useCollectionDocument` and `CollectionDocumentContext` for further testing

**Comment by 6thpath** [comment]
Hi @sadeghbarati, sorry for the late reply, there're more than `useCollectionDocument` and `CollectionDocumentContext` need to be exported, here's the StackBlitz you asked for

https://stackblitz.com/edit/vitejs-vite-bc1x7h

**Comment by sadeghbarati** [comment]
https://stackblitz.com/github/sadeghbarati/rac-select-multiple

@stephenh Hit the upload limit in Stackblitz so I create a GitHub repo for creating demo in Stackblitz
I have some errors in your demo even after exporting needed functions/components

**Comment by sadeghbarati** [comment]
@stephenh 

https://stackblitz.com/edit/github-qwxduv

I updated the manual package file with the latest changes in main branch, can you guide us here how to use that 
`select-multiple` component?

**Comment by Djiit** [comment]
Hey team, we'd love to see this implemented for our usage. Is there any way we can prio this ? Via a bounty or anything ?

**Comment by mryechkin** [comment]
> Combobox has no plans to support multi select.

Just wondering, is this still the team's stance on multi-select support within ComboBox in React Aria?

My experience with trying to implement this in user-land has been nothing but frustrations and headache.

**Comment by snowystinger** [comment]
At some point we'll tackle a TagField, which would fulfill that functionality. However, Combobox has no plans to support multi-select, see https://github.com/adobe/react-spectrum/issues/2140#issuecomment-2068999028 about the combobox aria pattern.

**Comment by mryechkin** [comment]
> At some point we'll tackle a TagField, which would fulfill that functionality. However, Combobox has no plans to support multi-select, see https://github.com/adobe/react-spectrum/issues/2140#issuecomment-2068999028 about the combobox aria pattern.

Thanks for the reply @snowystinger - I see. Hate to ask, but has the TagField been prioritized at all? Just wondering what kind of time frame we're looking at here. Appreciate the info!

**Comment by sadeghbarati** [comment]
I found these sources

- https://github.com/zaichaopan/tailwind-react-aria-components-starter-kits/blob/master/src/MultiSelect.tsx 
  - (doesn't have storybook demo)
  - `Combobox` + `TagsGroup`, the only bad thing about this component is that `Popover` will close after each item selection, @snowystinger  can we change this behavior manually by retouching the `Combobox` internal functions?
  
<br/>

- https://justd.co/docs/components/pickers/multi-select (wip)
- https://dotui.org/ (don't have multi-select demo just sharing xd)

**Comment by matthewdeutsch** [comment]
> Hate to ask, but has the TagField been prioritized at all?
We are waiting on our design team to redesign the component (scheduled likely for later this year) before engineering picks up the component for implementation

**Comment by romansndlr** [comment]
> > Hate to ask, but has the TagField been prioritized at all?
> 
> 
> 
> @mryechkin we are waiting on our design team to redesign the component (scheduled likely for later this year) before engineering picks up the component for implementation

Hi, is there maybe an update on this?

**Comment by matthewdeutsch** [comment]
@romansndlr our design team has picked up the TagField component from their backlog this month and has begun working through specifications, with designs to follow; in the meantime, our implementation is still on hold, but stay tuned 🤞

**Comment by romansndlr** [comment]
> [@romansndlr](https://github.com/romansndlr) our design team has picked up the TagField component from their backlog this month and has begun working through specifications, with designs to follow; in the meantime, our implementation is still on hold, but stay tuned 🤞

That's amazing! Thank you guys so much, would love it if y'all could keep us updated here on any advancements

**Comment by lsalling** [comment]
Hey, any updates on this one?

**Comment by eashish93** [comment]
This is full implementation of multi-select using claude sonnet 3.7 (thinking). I did it in almost 30+ prompts. But working correctly.
```code
import {
  Button,
  Label,
  ListBox,
  ListBoxItem,
  Popover,
  FieldError,
  Text,
  Input,
  ComboBox,
  MenuTrigger,
} from 'react-aria-components';
import type {
  ListBoxItemProps,
  ValidationResult,
  Key,
  Selection,
  ComboBoxProps as AriaComboBoxProps,
} from 'react-aria-components';
import { cx } from 'cva';
import React, { forwardRef, type ForwardedRef, useState, useRef, useEffect } from 'react';

// Simple down arrow icon
const IconChevronDown = ({ size = 16 }) => (
  <svg 
    width={size} 
    height={size} 
    viewBox="0 0 16 16" 
    fill="none" 
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M4 6L8 10L12 6" stroke="currentColor" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round"/>
  </svg>
);

// Simple X icon for removing tags
const IconX = ({ size = 12 }) => (
  <svg 
    width={size} 
    height={size} 
    viewBox="0 0 12 12" 
    fill="none" 
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M9 3L3 9M3 3L9 9" stroke="currentColor" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round"/>
  </svg>
);

export interface MultiSelectProps<T extends object> {
  label?: string;
  labelClassName?: string;
  children: React.ReactNode;
  helpText?: string | React.ReactNode;
  errorMessage?: string | ((v: ValidationResult) => string);
  items?: Iterable<T>;
  className?: string;
  isDisabled?: boolean;
  isRequired?: boolean;
  isInvalid?: boolean;
  placeholder?: string;
  selectedKeys?: Selection;
  defaultSelectedKeys?: Selection;
  onSelectionChange?: (keys: Selection) => void;
}

function AriaMultiSelect<T extends object>(
  {
    label,
    labelClassName,
    children,
    items,
    helpText,
    errorMessage,
    className,
    isDisabled,
    isInvalid,
    isRequired,
    placeholder = "Select options",
    selectedKeys: externalSelectedKeys,
    defaultSelectedKeys,
    onSelectionChange: externalOnSelectionChange,
    ...props
  }: MultiSelectProps<T>,
  ref: ForwardedRef<HTMLDivElement>
) {
  // Local selection state if not controlled externally
  const [localSelectedKeys, setLocalSelectedKeys] = useState<Selection>(
    defaultSelectedKeys || new Set([])
  );
  
  // For input value in ComboBox
  const [inputValue, setInputValue] = useState('');
  
  // Use controlled or uncontrolled selection
  const selectedKeys = externalSelectedKeys !== undefined ? externalSelectedKeys : localSelectedKeys;
  
  // Handle selection change
  const handleSelectionChange = (key: Key | null) => {
    if (key && typeof selectedKeys !== 'string') {
      // Create a new set with the existing keys
      const newKeys = new Set(selectedKeys);
      
      // Add the key if it doesn't exist, otherwise remove it
      if (newKeys.has(key)) {
        newKeys.delete(key);
      } else {
        newKeys.add(key);
      }
      
      // Update selection
      if (externalOnSelectionChange) {
        externalOnSelectionChange(newKeys);
      } else {
        setLocalSelectedKeys(newKeys);
      }
      
      // Clear input value after selection
      setInputValue('');
    }
  };
  
  // Reference to the input wrapper for positioning the popover
  const inputWrapperRef = useRef<HTMLDivElement>(null);
  
  // Reference to the input element for focus management
  const inputRef = useRef<HTMLInputElement>(null);
  
  // Filter function for ComboBox that filters by input text and excludes selected items
  const filterItems = (textValue: string, inputValue: string) => {
    return textValue.toLowerCase().includes(inputValue.toLowerCase());
  };
  
  // Track the open state of the dropdown
  const [isOpen, setIsOpen] = useState(false);
  
  // Track whether user is in the middle of removing an item
  const [isRemoving, setIsRemoving] = useState(false);
  
  // Convert selection to array for easier handling
  const selectedKeysArray = typeof selectedKeys !== 'string' ? Array.from(selectedKeys || []) : [];
  
  // Remove a selected item
  const removeItem = (key: Key) => {
    if (typeof selectedKeys !== 'string') {
      const newKeys = new Set(selectedKeys);
      newKeys.delete(key);
      
      if (externalOnSelectionChange) {
        externalOnSelectionChange(newKeys);
      } else {
        setLocalSelectedKeys(newKeys);
      }
    }
  };
  
  // This special effect ensures the dropdown stays open when we want it to
  useEffect(() => {
    if (isRemoving && !isOpen) {
      setIsOpen(true);
    }
  }, [isOpen, isRemoving]);
  
  // Get display text for a key
  const getDisplayText = (key: Key): string => {
    // Try to find from ListBoxItem elements
    const allChildren = React.Children.toArray(children);
    
    const result = allChildren.find((child) => {
      if (React.isValidElement(child)) {
        const childProps = child.props as any;
        const childKey = child.key?.toString().replace(/^\$/, '') || childProps.id || childProps.value;
        return String(childKey) === String(key);
      }
      return false;
    });
    
    if (result && React.isValidElement(result)) {
      const props = result.props as any;
      if (props.textValue) return props.textValue;
      if (typeof props.children === 'string') return props.children;
    }
    
    // Try direct text matching for common items
    if (typeof key === 'string') {
      // Match text directly with the key
      const textDirectMatch = allChildren.find(child => {
        if (React.isValidElement(child)) {
          const props = child.props as any;
          if (typeof props.children === 'string' && props.children.toLowerCase() === key.toLowerCase()) {
            return true;
          }
          if (props.textValue && props.textValue.toLowerCase() === key.toLowerCase()) {
            return true;
          }
        }
        return false;
      });
      
      if (textDirectMatch && React.isValidElement(textDirectMatch)) {
        const props = textDirectMatch.props as any;
        return typeof props.children === 'string' ? props.children : props.textValue;
      }
    }
    
    // If not found, return the key itself
    return String(key);
  };
  
  // Handle keyboard events
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    // If Backspace is pressed while input is empty, remove the last selected item
    if (e.key === 'Backspace' && inputValue === '' && typeof selectedKeys !== 'string') {
      const keysArray = Array.from(selectedKeys);
      if (keysArray.length > 0) {
        const lastKey = keysArray[keysArray.length - 1];
        removeItem(lastKey);
        // Prevent default behavior
        e.preventDefault();
        e.stopPropagation();
      }
    }
  };
  
  // For forwarding the ref
  const combinedRef = (node: HTMLDivElement) => {
    if (typeof ref === 'function') {
      ref(node);
    } else if (ref) {
      ref.current = node;
    }
  };
  
  return (
    <div ref={combinedRef} className={className}>
      {label && (
        <Label
          className={cx('block mb-1 font-medium', isInvalid && 'text-red-500', labelClassName)}>
          {label}
          {isRequired && <span className="text-red-500 ml-1">*</span>}
        </Label>
      )}

      <ComboBox
        inputValue={inputValue}
        onInputChange={setInputValue}
        onSelectionChange={handleSelectionChange}
        menuTrigger="focus"
        items={items}
        isDisabled={isDisabled}
        aria-label={label}
        selectedKey={null} // Force no single selection
        disabledKeys={selectedKeysArray} // Disable already selected items
        defaultFilter={filterItems} // Apply text filtering
        onOpenChange={setIsOpen}
      >
        <div
          ref={inputWrapperRef}
          className={cx(
            'flex items-center border rounded min-h-[56px]',
            isInvalid ? 'border-red-500' : 'border-gray-300',
            isDisabled && 'opacity-50 cursor-not-allowed'
          )}>
          <div className="flex flex-grow flex-wrap gap-1 items-center pl-3">
            {selectedKeys &&
              typeof selectedKeys !== 'string' &&
              Array.from(selectedKeys).length > 0 && (
                <div className="flex flex-wrap gap-1 my-1 mr-1">
                  {Array.from(selectedKeys).map((key) => (
                    <span
                      key={String(key)}
                      className="bg-gray-100 rounded flex items-center gap-1 px-2 py-1 text-sm">
                      {getDisplayText(key)}
                      <div
                        role="button"
                        tabIndex={0}
                        aria-label={`Remove ${getDisplayText(key)}`}
                        className="hover:bg-gray-200 rounded-full p-0.5 cursor-pointer"
                        onMouseDown={(e) => {
                          // Prevent default to avoid losing focus
                          e.preventDefault();
                          e.stopPropagation();
                        }}
                        onClick={(e) => {
                          removeItem(key);
                          // Prevent default to avoid dropdown closing
                          e.preventDefault();
                          e.stopPropagation();
                          // Focus back on the input field
                          inputRef.current?.focus();
                        }}
                        onKeyDown={(e) => {
                          if (e.key === 'Enter' || e.key === ' ') {
                            removeItem(key);
                            // Prevent default to avoid dropdown closing
                            e.preventDefault();
                            e.stopPropagation();
                            // Focus back on the input field
                            inputRef.current?.focus();
                          }
                        }}>
                        <IconX />
                      </div>
                    </span>
                  ))}
                </div>
              )}
            <Input
              ref={inputRef}
              className="flex-1 min-w-[50px] outline-none p-2 rounded-sm border-none"
              placeholder={
                selectedKeys &&
                typeof selectedKeys !== 'string' &&
                Array.from(selectedKeys).length > 0
                  ? ''
                  : placeholder
              }
              onKeyDown={handleKeyDown}
            />
          </div>
          <Button className="px-3 appearance-none outline-none bg-transparent">
            <IconChevronDown aria-hidden="true" />
          </Button>
        </div>

        <Popover
          containerPadding={0}
          triggerRef={inputWrapperRef}
          className={cx(
            'min-w-[--trigger-width] py-2 overflow-auto rounded-md border border-on-secondary-container/[0.08] shadow-sm bg-white',
            'entering:animate-in entering:fade-in exiting:animate-out exiting:fade-out'
          )}
          // Fix: https://github.com/adobe/react-spectrum/issues/5331 (popover width inside modal is wrong)
          style={{
            width: inputWrapperRef.current
              ? `${inputWrapperRef.current.getBoundingClientRect().width}px`
              : undefined,
          }}>
          <ListBox
            items={items}
            selectionMode="multiple"
            selectedKeys={selectedKeys}
            disabledKeys={selectedKeysArray}
            onSelectionChange={(keys) => {
              // Do nothing - we handle this through ComboBox's onSelectionChange
              // This prevents double selection handling
            }}
            disallowEmptySelection={false}
            className="p-1 outline-none"
            autoFocus="first">
            {children}
          </ListBox>
        </Popover>
      </ComboBox>

      {helpText && !isInvalid && <div className="text-sm text-gray-500 mt-1">{helpText}</div>}

      {isInvalid && <FieldError className="text-sm text-red-500 mt-1">{errorMessage}</FieldError>}
    </div>
  );
}

// ListBoxItem implementation
function AriaListBoxItem(
  { className, ...props }: ListBoxItemProps,
  ref: ForwardedRef<HTMLDivElement>
) {
  return (
    <ListBoxItem
      {...props}
      ref={ref}
      className={cx(
        "px-3 py-2 rounded cursor-pointer outline-none",
        "hover:bg-gray-100",
        "selected:bg-gray-100 selected:font-medium",
        "focus:bg-gray-100",
        "disabled:opacity-50 disabled:cursor-default disabled:hover:bg-transparent disabled:text-gray-400",
        className
      )}
    />
  );
}

const _AriaListBoxItem = forwardRef(AriaListBoxItem) as <T>(
  props: ListBoxItemProps<T> & { ref?: React.ForwardedRef<HTMLDivElement> }
) => React.ReactElement;

const _MultiSelect = forwardRef(AriaMultiSelect) as <T extends object>(
  props: MultiSelectProps<T> & { ref?: React.ForwardedRef<HTMLDivElement> }
) => React.ReactElement;

export { _MultiSelect as MultiSelect, _AriaListBoxItem as ListBoxItem };


```

**Comment by lixiaoyan** [comment]
> 1. Assistive technology wouldn't know how to announce multiple selection in a combobox.

Hi @snowystinger, is it possible to create a searchable, accessible multi-select that functions similarly to a multi-select combobox?

**Comment by muhammadhaffad** [comment]
This is my custom implementation of a multi-combobox, combined with `<select multiple />`.

Feedback and corrections are welcome.

Reference combobox: https://stackblitz.com/edit/vitejs-vite-iycf5n

```
import { useFormValidation } from "@react-aria/form";
import { useFormValidationState } from "@react-stately/form";
import { useRef, useState } from "react";
import { FormContext, Group, TableBody, TagList, useSlottedContext } from "react-aria-components";
import { useListData } from "react-stately";
import { ComboBox, ComboBoxItem } from "@/components/ui/ComboBox";
import { Tag, TagGroup } from "@/components/ui/TagGroup";

function MultiComboBox({ items, selectedItems, isRequired = false, name = '', ...props }) {
    const selectRef = useRef();
    const formContext = useSlottedContext(FormContext) || {};
    const validationState = useFormValidationState({
        isRequired: isRequired,
        name: name,
        validationBehavior: formContext.validationBehavior,
        value: selectedItems
    });
    useFormValidation({
		validationBehavior: formContext.validationBehavior,
		focus: () => selectRef.current?.focus()
	}, validationState, selectRef);
    const list = useListData({
        initialItems: selectedItems,
    });
    const availableList = useListData({
        initialItems: items.filter((item) => !selectedItems.map((item) => item.id).includes(item.id)),
    });
    return (
        <Group>
            <select
                className="hidden"
                multiple
                value={list.items.map((item) => item.id)}
                onChange={() => { }}
                required={isRequired}
                name={name}
                ref={selectRef}
            >
                {items.map((item) => (
                    <option key={item.id} value={item.id}>
                        {item.name}
                    </option>
                ))}
            </select>
            <TagGroup onRemove={(keys) => {
                let key = keys.values().next().value;
                availableList.append(list.getItem(key));
                list.remove(key);
                validationState.commitValidation();
            }}>
                <TagList items={list.items}>
                    {(item) => <Tag id={item.id}>{item.name}</Tag>}
                </TagList>
            </TagGroup>
            <ComboBox
                items={availableList.items}
                renderEmptyState={() => <p className="text-center p-2">No available permissions</p>}
                onSelectionChange={(key) => {
                    if (!availableList.getItem(key)) {
                        return;
                    }
                    list.append(availableList.getItem(key));
                    availableList.remove(key);
                    validationState.commitValidation();
                    console.info(validationState);
                }}
                selectedKey={null}
                allowsEmptyCollection
                isInvalid={validationState.displayValidation.isInvalid}
                errorMessage={validationState.displayValidation.validationErrors.join(', ')}
                {...props}
            >
                {(item) => <ComboBoxItem id={item.id}>{item.name}</ComboBoxItem>}
            </ComboBox>
        </Group>
    );
}
```

**Comment by devongovett** [comment]
In React Aria Components v1.13.0 released today, we have support for multiple selection in the Select component! See https://react-spectrum.adobe.com/react-aria/Select.html#multiple-selection.

This can be combined with TagGroup and Autocomplete to create something similar to a multi-select ComboBox: https://stackblitz.com/edit/rac-select-taggroup?file=src%2FExample.tsx

We do still plan to introduce a TagField component in the future which will be like a TagGroup but with inline editing similar to the "To" field in gmail, but the multiple Select component should cover a good number of use cases as well.

**Comment by berkerdemirer** [comment]
I have one question regarding the proposed solution. When page is navigated by using tab and VoiceOver, the first focus is on the TagGroup. How does the user will know they are actually inside of a Select? If only they tab and focus to the plus button they can hear that its a Select.

**Comment by snowystinger** [comment]
It announces "Selected states, group" and the next tab announces "Selected States, states, listbox popup collapsed"
So I think it's not a big leap to associate the two elements. However, you could add an element around the whole thing with role="group" and add information there, or it could be added to just the aria-label for the TagGroup itself.

What are you expecting to be announced though? Or what would be announced in your ideal world? Maybe we can offer better suggestions knowing that.

**Comment by berkerdemirer** [comment]
I will try out the group solution. My main concern was that when users are in "selected states, group" they might not know that it is a select field component. So, there is no hint of what next tab might be. IMO, it should be announced that the user is currently inside of a select field, viewing the selected items and that next tab focus will be the select field's trigger. Hard to put into words but that's the idea.

**Comment by lixiaoyan** [comment]
I still think that we should allow multiple selection in comboboxes. It already exists in a native input.
https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/multiple#email_input

**Comment by berkerdemirer** [comment]
In the end, i ended up using downshiftjs and created my own component
